### ISSUE TITLE
Global font size makes it difficult to re-use components with other frameworks

### ISSUE THREAD

**Issue by adborden** [18F/fec-style#286]
FEC style defines a reduced [62.5% font size](https://github.com/18F/fec-style/blob/master/scss/elements/_typography.scss#L15) and then uses a lot of `rem` units for padding, margin, height, font-size in components like the disclaimer and masthead.

When integrating the FEC styles with [eregs](https://github.com/18F/fec-eregs), the global font size is 100%, which really throws off some of the component sizes using `rem`s. In this particular case, reducing the global font-size has too many undesired consequences.

I'm overriding the styles of the FEC components using fixed sizes (px), but I wonder if using a different unit (`em`) for some of these components would be more appropriate, since things like the masthead and disclaimer are probably designed to be a very particular size, no matter what the rest of the site is scaled to.

Wondering if you folks have any thoughts on this kind of problem.

**Comment by adborden** [comment]
The [Frontend guild's CSS styleguide](https://pages.18f.gov/frontend/css-coding-styleguide/units/) suggests setting the global font-size to 10px.

**Comment by noahmanger** [comment]
Good points! That should be fine. Just asking in #g-frontend to see what others think.

**Comment by noahmanger** [comment]
I opened up an issue for discussion in the front-end guide: https://github.com/18F/frontend/issues/118

**Comment by noahmanger** [comment]
So, the conversation over at the [web design standards on this issue](https://github.com/18F/web-design-standards/issues/992) is coming down on the side of setting the base font-size to a percentage. 

As for whether that percentage is 100% or 62.5% only matters in as much as it affects a given situation (like this one). 

I'm not opposed to changing fec-style to 100% base font-size, but then it would just entail updating probably hundreds of REM declarations to use the new math. I'm not familiar with eregs or what you're trying to do there, so maybe there's another solution?

**Comment by noahmanger** [comment]
Is there anything further you want to do on this, or can we icebox until we decide which method we're going to use for eregs?

**Comment by adborden** [comment]
@noahmanger I'm following up on this now as part of looking at how to integrate `fec-style` with `fec-eregs`. Looking at some other frameworks to see what strategies they're using. I would leave this open for a little longer.

**Comment by adborden** [comment]
Okay, finally spent some time this sprint on this problem.

tl;dr, I think instead of using fixed sizes everywhere, we should use a function which calculates the size for us. e.g.

```scss
$base-size-scale-factor: 10; // This value would be different for betaFEC than for eregs
@function u($size) {
  // magic splitting of $size into components, any unit e.g. px, rem, etc.
  // magic scaling of each component based on $base-size-scale-factor;
}

.callout {
  @include clearfix();
  padding: u(1rem 1.5rem);
  margin-bottom: u(2rem);
  border-radius: 4px;
}
```

Luckily, the magic parts have actually been done for us in the following links. Details...

So, it's not so much what the global size is set, to but that we're just using flat out `rem`s for our component sizing. Really, we should either be using a function or a variable to scale from.

For example, [Bootstrap](https://github.com/twbs/bootstrap/blob/master/less/variables.less#L51) set's a `base-font-size` for all of its font-size scaling calculations. While they don't explicitly use this for component sizing, they do [use base sizing variables](https://github.com/twbs/bootstrap/blob/master/less/variables.less#L88) for some components which are based on the "idea" of a base-font-size. If we were using a base-size variable, I could change this to the value of ereg's rem, and then all the FEC components would scale appropriately.

Now, consider using a function for calculating the sizes. [A lot of folks](https://css-tricks.com/snippets/css/less-mixin-for-rem-font-sizing/) are using mixins for `rem` to `px` fallbacks. But I found [this article](https://davidwalsh.name/rem-px-browser-function-sass) uses a function which dynamically returns the fallback based on a variable. It then generates two stylesheets, one with `rem` and one fallback with all units as `px`.

It would be sooo cool if Sass had better introspection, where I could just on the fly write override rules that looked at whatever font-size/padding/margin was set and recalculate it on the fly. Sadly, this does not exist. Libsass supports plugins, but I see very few plugins actually being built.

What do ya think?

**Comment by noahmanger** [comment]
Very interesting idea. I've definitely seen other things like this used. I think my only concern is that it will be pretty tricky to make this replacement throughout the entire range of files, but it's probably the best option. Could we do a global find-and-replace using some fancy regex? (Regex is still dark sorcery to me, tbh).

**Comment by adborden** [comment]
I was playing with this strategy on Friday and came up with this sed script to do the regex. It works with GNU sed, not so sure about BSD sed which would be on mac, but we'd only have to run it once anyway.

```sed
# A comment, but jump to `a` to process it
\:rem.*//: b a

# A comment, skip the line.
\://: { p ; d }

: a 

# Wrap the line with the function. i.e. u(1rem 0 10px 0); Matches include:
# font-size: 1rem;
# padding-right: -1rem;
# margin-right: -.1rem;
# margin-right: -12.1rem;
# margin: 1rem 0 0 0;
# margin: 1rem 0 10px 0;
/rem/ s/:\s*((-?([0-9]+)?(\.?[0-9]+)(\w+|%)?\s*)+)\s*;/: u(\1);/
```

The overall approach seems to work well in fact. Using the `rem` -> `px` conversion with a size of `10pz` seems to work well. Our breakpoints are in `em`s so I'm looking at that now :skull:

**Comment by noahmanger** [comment]
Aw yeah. That's great. Just checking, this will catch all properties, not just margin and padding?

**Comment by adborden** [comment]
Uh, well, it will catch any property that uses single values (`1rem`), or multiple values (`1rem 0 0 2rem`) the way margin/padding do. I didn't want to catch _every_ instance of `rem` because the conversion isn't globally applicable. For example, I don't think we want to change border sizes. The above regex will not match `border: 1rem solid black;`, but it actually would catch `border: 1rem;` (maybe I should fix the regex to exclude this). Our `border` properties are usually written with `px` anyway so I didn't notice this as an issue when reviewing the converted files. We'll have to review the diff carefully once we're ready.

In the event of a mistake (we wrap something accidentally), it won't affect betaFEC, because the size function effectively does a scaling of 1, so no visible change. You would only notice it on eregs and we could just treat these as one-off bugs and can be fixed manually. Or in case we missed an entire class of issues, we could write another regex.

**Comment by noahmanger** [comment]
Nice. Yeah, there will also probably be some `background-size` properties set as part of `background:` that will need some individual attention, but this sounds pretty straightforward.

**Comment by adborden** [comment]
This is also relevant, bourbon includes a [modular-scaling function](http://bourbon.io/docs/#modular-scale) for text and also a [global em-base variable](http://bourbon.io/docs/#em-base) for it's `px-to-em`/`px-to-rem` conversion.

**Comment by noahmanger** [comment]
Resolved with #369
